File: Beecrowd_Parte5/Sorvete_3069.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Scanner;

public class Sorvete_3069 {

    static class Interval {
        int start;
        int end;

        Interval(int start, int end) {
            this.start = start;
            this.end = end;
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int testNumber = 1;

        while (true) {
            int P = sc.nextInt(); // Comprimento da praia
            int S = sc.nextInt(); // Número de sorveteiros

            if (P == 0 && S == 0) break;

            ArrayList<Interval> intervals = new ArrayList<>();

            for (int i = 0; i < S; i++) {
                int U = sc.nextInt();
                int V = sc.nextInt();
                intervals.add(new Interval(U, V));
            }

            // Ordena os intervalos pela coordenada de início
            Collections.sort(intervals, new Comparator<Interval>() {
                public int compare(Interval a, Interval b) {
                    return Integer.compare(a.start, b.start);
                }
            });

            // Mescla os intervalos
            ArrayList<Interval> mergedIntervals = new ArrayList<>();
            if (!intervals.isEmpty()) {
                Interval current = intervals.get(0);
                for (int i = 1; i < intervals.size(); i++) {
                    Interval next = intervals.get(i);
                    if (current.end >= next.start) {
                        // Atualiza o intervalo atual
                        current.end = Math.max(current.end, next.end);
                    } else {
                        // Adiciona o intervalo mesclado à lista final
                        mergedIntervals.add(current);
                        current = next;
                    }
                }
                // Adiciona o último intervalo mesclado
                mergedIntervals.add(current);
            }

            // Imprime o resultado do teste
            System.out.println("Teste " + testNumber++);
            for (Interval interval : mergedIntervals) {
                System.out.println(interval.start + " " + interval.end);
            }
            System.out.println(); // linha em branco entre testes
        }

        sc.close();
    }
}

File: Beecrowd_Parte5/Sorvete_3069.c
#include <stdio.h>
#include <stdlib.h>

// Estrutura para representar um intervalo [start, end]
typedef struct {
    int start;
    int end;
} Interval;

// Função para comparar dois intervalos pelo início (start)
int compare(const void *a, const void *b) {
    Interval *intervalA = (Interval *)a;
    Interval *intervalB = (Interval *)b;
    return intervalA->start - intervalB->start;
}

int main() {
    int testNumber = 1;
    int P, S;

    while (1) {
        scanf("%d %d", &P, &S);

        if (P == 0 && S == 0) break;

        // Leitura dos intervalos de trabalho de cada sorveteiro
        Interval *intervals = (Interval *)malloc(S * sizeof(Interval));
        for (int i = 0; i < S; i++) {
            scanf("%d %d", &intervals[i].start, &intervals[i].end);
        }

        // Ordenação dos intervalos pelo início (start)
        qsort(intervals, S, sizeof(Interval), compare);

        // Mesclagem dos intervalos
        Interval *mergedIntervals = (Interval *)malloc(S * sizeof(Interval));
        int mergedCount = 0;

        if (S > 0) {
            mergedIntervals[0] = intervals[0];
            mergedCount = 1;

            for (int i = 1; i < S; i++) {
                if (intervals[i].start <= mergedIntervals[mergedCount - 1].end) {
                    // Atualiza o fim do último intervalo mesclado se houver sobreposição
                    mergedIntervals[mergedCount - 1].end = intervals[i].end > mergedIntervals[mergedCount - 1].end
                                                            ? intervals[i].end
                                                            : mergedIntervals[mergedCount - 1].end;
                } else {
                    // Adiciona um novo intervalo mesclado
                    mergedIntervals[mergedCount] = intervals[i];
                    mergedCount++;
                }
            }
        }

        // Impressão do resultado para o conjunto de teste
        printf("Teste %d\n", testNumber++);
        for (int i = 0; i < mergedCount; i++) {
            printf("%d %d\n", mergedIntervals[i].start, mergedIntervals[i].end);
        }
        printf("\n");

        // Liberação de memória
        free(intervals);
        free(mergedIntervals);
    }

    return 0;
}
